apps/emqx/src/emqx_socket_connection.erl

```diff
@@ -0,0 +1,1267 @@
+            process_msgs(NextMsgs, NState);
+        {ok, NextMsg, NState} ->
+            process_msg(NextMsg, NState);
+        {stop, Reason, NState} ->
+            {stop, Reason, NState};
+        {stop, Reason} ->
+            {stop, Reason, State}
+    catch
+        exit:normal ->
+            {stop, normal, State};
+        exit:shutdown ->
+            {stop, shutdown, State};
+        exit:{shutdown, _} = Shutdown ->
+            {stop, Shutdown, State};
+        Exception:Context:Stack ->
+            {stop,
+                #{
+                    exception => Exception,
+                    context => Context,
+                    stacktrace => Stack
+                },
+                State}
+    end.
+
+%%--------------------------------------------------------------------
+%% Handle a Msg
+handle_msg({'$gen_call', From, Req}, State) ->
+    case handle_call(From, Req, State) of
+        {reply, Reply, NState} ->
+            gen_server:reply(From, Reply),
+            {ok, NState};
+        {stop, Reason, Reply, NState} ->
+            gen_server:reply(From, Reply),
+            stop(Reason, NState)
+    end;
+handle_msg({'$gen_cast', Req}, State) ->
+    NewState = handle_cast(Req, State),
+    {ok, NewState};
+handle_msg({'$socket', Socket, select, Handle}, State = #state{sockstate = SS}) ->
+    case SS of
+        idle ->
+            handle_data_ready(Socket, State);
+        #congested{handle = Handle} ->
+            handle_send_ready(Socket, SS, State);
+        _ ->
+            handle_data_ready(Socket, State)
+    end;
+handle_msg({'$socket', _Socket, abort, {_Handle, Reason}}, State = #state{sockstate = SS}) ->
+    case SS =/= closed of
+        true ->
+            handle_info({sock_error, Reason}, State);
+        false ->
+            %% In case there were more than 1 outstanding select:
+            {ok, State}
+    end;
+handle_msg({recv, Data}, State) ->
+    handle_data(Data, false, State);
+handle_msg({recv_more, Data}, State) ->
+    handle_data(Data, true, State);
+handle_msg({incoming, Packet}, State) ->
+    ?TRACE("MQTT", "mqtt_packet_received", #{packet => Packet}),
+    handle_incoming(Packet, State);
+handle_msg({outgoing, Packets}, State) ->
+    handle_outgoing(Packets, State);
+handle_msg(
+    Deliver = {deliver, _Topic, _Msg},
+    #state{gc_tracker = {ActiveN, _, _}} = State
+) ->
+    ?BROKER_INSTR_SETMARK(t0_deliver, {_Msg#message.extra, ?BROKER_INSTR_TS()}),
+    Delivers = [Deliver | emqx_utils:drain_deliver(ActiveN)],
+    with_channel(handle_deliver, [Delivers], State);
+handle_msg({connack, ConnAck}, State) ->
+    handle_outgoing(ConnAck, State);
+handle_msg({close, Reason}, State) ->
+    %% @FIXME here it could be close due to appl error.
+    ?TRACE("SOCKET", "socket_force_closed", #{reason => Reason}),
+    handle_info({sock_closed, Reason}, close_socket(State));
+handle_msg({event, connected}, State = #state{channel = Channel}) ->
```

Comment:
this module seems to lack of the trigger for `{event, connected | disconnected | running}`